src/librustc_macros/src/query.rs

```diff
@@ -0,0 +1,320 @@
+use proc_macro::TokenStream;
+use proc_macro2::Span;
+use syn::{
+    Token, Ident, Type, Attribute, ReturnType, Expr, Block,
+    braced, parenthesized, parse_macro_input,
+};
+use syn::parse::{Result, Parse, ParseStream};
+use syn::punctuated::Punctuated;
+use quote::quote;
+use crate::tt::TS;
+
+struct IdentOrWild(Ident);
+
+impl Parse for IdentOrWild {
+    fn parse(input: ParseStream<'_>) -> Result<Self> {
+        Ok(if input.peek(Token![_]) {
+            input.parse::<Token![_]>()?;
+            IdentOrWild(Ident::new("_", Span::call_site()))
+        } else {
+            IdentOrWild(input.parse()?)
+        })
+    }
+}
+
+enum QueryAttribute {
+    Desc(Option<Ident>, Punctuated<Expr, Token![,]>),
+    Cache(Option<Ident>, Expr),
+    LoadCached(Ident, Ident, Block),
+    FatalCycle,
+}
+
+impl Parse for QueryAttribute {
+    fn parse(input: ParseStream<'_>) -> Result<Self> {
+        let attr: Ident = input.parse()?;
+        if attr == "desc" {
+            let attr_content;
+            braced!(attr_content in input);
+            let tcx = if attr_content.peek(Token![|]) {
+                attr_content.parse::<Token![|]>()?;
+                let tcx = attr_content.parse()?;
+                attr_content.parse::<Token![|]>()?;
+                Some(tcx)
+            } else {
+                None
+            };
+            let desc = attr_content.parse_terminated(Expr::parse)?;
+            if !attr_content.is_empty() {
+                panic!("unexpected tokens in block");
+            };
+            Ok(QueryAttribute::Desc(tcx, desc))
+        } else if attr == "cache" {
+            let attr_content;
+            braced!(attr_content in input);
+            let tcx = if attr_content.peek(Token![|]) {
+                attr_content.parse::<Token![|]>()?;
+                let tcx = attr_content.parse()?;
+                attr_content.parse::<Token![|]>()?;
+                Some(tcx)
+            } else {
+                None
+            };
+            let expr = attr_content.parse()?;
+            if !attr_content.is_empty() {
+                panic!("unexpected tokens in block");
+            };
+            Ok(QueryAttribute::Cache(tcx, expr))
+        } else if attr == "load_cached" {
+            let args;
+            parenthesized!(args in input);
+            let tcx = args.parse()?;
+            args.parse::<Token![,]>()?;
+            let id = args.parse()?;
+            if !args.is_empty() {
+                panic!("unexpected tokens in arguments");
+            };
+            let block = input.parse()?;
+            Ok(QueryAttribute::LoadCached(tcx, id, block))
+        } else if attr == "fatal_cycle" {
+            Ok(QueryAttribute::FatalCycle)
+        } else {
+            panic!("unknown query modifier {}", attr)
```

Comment:
Here (and elsewhere) I would prefer to see the parsing code return a syn::Error rather than panic. That way the error can be correctly spanned to show which part of a potentially large macro input was problematic.